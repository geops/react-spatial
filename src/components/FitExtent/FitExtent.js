import React from 'react';
import PropTypes from 'prop-types';
import { FaExpand } from 'react-icons/fa';
import OLMap from 'ol/Map';
import Button from '../Button';

const propTypes = {
  /**
   * An ol map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * The extent to be zoomed.
   */
  extent: PropTypes.arrayOf(PropTypes.number).isRequired,

  /**
   * Title for the fitExtent button.
   */
  title: PropTypes.string,

  /**
   * CSS class of the fitExtent button.
   */
  className: PropTypes.string,

  /**
   * Children content of the fitExtent button.
   */
  children: PropTypes.node,
};

const defaultProps = {
  title: 'Fit Extent',
  className: 'tm-button tm-round-blue',
  children: <FaExpand focusable={false} />,
};

/**
 * This component creates a button to zoom to the given extent.
 */
function FitExtent({ map, extent, title, className, children }) {
  return (
    <Button
      className={className}
      title={title}
      onClick={() => {
        map.getView().cancelAnimations();
        map.getView().fit(extent, map.getSize());
      }}
    >
      {children}
    </Button>
  );
}

FitExtent.propTypes = propTypes;
FitExtent.defaultProps = defaultProps;

export default FitExtent;
