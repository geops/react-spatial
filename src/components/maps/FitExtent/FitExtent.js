import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import OLMap from 'ol/Map';

const propTypes = {
  /**
   * An ol map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * The extent to be zoomed.
   */
  extent: PropTypes.arrayOf(PropTypes.number).isRequired,

  /**
   * CSS class  for the fitExtent button.
   */
  className: PropTypes.string,

  /**
   * Button content.
   */
  children: PropTypes.node.isRequired,
};

const defaultProps = {
  className: 'rs-fit-extent',
};

/**
 * This component creates a button to zoom to the given extent.
 */
function FitExtent({ map, extent, className, children, ...other }) {
  const fit = useCallback((evt) => {
    if (evt.which && evt.which !== 13) {
      return;
    }
    map.getView().cancelAnimations();
    map.getView().fit(extent, map.getSize());
  });

  return (
    <div
      className={className}
      role="button"
      tabIndex="0"
      onClick={fit}
      onKeyPress={fit}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...other}
    >
      {children}
    </div>
  );
}

FitExtent.propTypes = propTypes;
FitExtent.defaultProps = defaultProps;

export default FitExtent;
