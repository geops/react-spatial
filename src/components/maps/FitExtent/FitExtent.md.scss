.rs-fit-extent {
  background: #f00;
  border-radius: 50%;
  color: white;
  height: 50px;
  width: 50px;
  display: flex;
  justify-content: center;
  align-items: center;
}
