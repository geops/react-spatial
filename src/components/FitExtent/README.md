#

This demonstrates the use of FitExtent.

```jsx
import React from 'react';
import FitExtent from 'react-spatial/components/FitExtent';
import BasicMap from 'react-spatial/components/BasicMap';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData);
const extent = [-15380353.1391, 2230738.2886, -6496535.908, 6927029.2369];

function FitExtentExample() {
  return (
    <div className="tm-fitextent-example">
      <BasicMap map={map} layers={layers} />
      <FitExtent map={map} extent={extent} />
    </div>
  );
};

<FitExtentExample />;
```
