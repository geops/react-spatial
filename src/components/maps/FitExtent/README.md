#

This demonstrates the use of FitExtent.

```jsx
import React from 'react';
import FitExtent from 'react-spatial/components/maps/FitExtent';
import BasicMap from 'react-spatial/components/maps/BasicMap';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

const map = new OLMap({ controls: [] });
const layers = ConfigReader.readConfig(treeData);
const extent = [-15380353.1391, 2230738.2886, -6496535.908, 6927029.2369];

function FitExtentExample() {
  return (
    <div>
      <BasicMap map={map} layers={layers} tabIndex={0} />
      <FitExtent map={map} extent={extent}>fit!</FitExtent>
    </div>
  );
};

<FitExtentExample />;
```
