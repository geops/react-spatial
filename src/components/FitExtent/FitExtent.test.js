import Adapter from "@cfaester/enzyme-adapter-react-18";
import { configure, shallow } from "enzyme";
import OLMap from "ol/Map";
import OLView from "ol/View";
import React from "react";
import renderer from "react-test-renderer";

import FitExtent from "./FitExtent";

configure({ adapter: new Adapter() });

const extent = [1, 2, 3, 4];

test("Button should match snapshot.", () => {
  const map = new OLMap({});
  const component = renderer.create(
    <FitExtent extent={extent} map={map}>
      FitExtent
    </FitExtent>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Should fit the extent.", () => {
  const map = new OLMap({ view: new OLView({ center: [0, 0], zoom: 7 }) });
  const wrapper = shallow(
    <FitExtent className="fit-ext" extent={extent} map={map}>
      FitExtent
    </FitExtent>,
  );
  wrapper.find(".fit-ext").first().simulate("click", {});
  const calculatedExtent = map.getView().calculateExtent(map.getSize());

  expect(calculatedExtent).toStrictEqual([1, 2, 3, 4]);
});

test("Should fit the extent on return.", () => {
  const map = new OLMap({ view: new OLView({ center: [0, 0], zoom: 7 }) });
  const wrapper = shallow(
    <FitExtent className="fit-ext" extent={extent} map={map}>
      FitExtent
    </FitExtent>,
  );
  wrapper.find(".fit-ext").first().simulate("click", { which: 13 });
  const calculatedExtent = map.getView().calculateExtent(map.getSize());

  expect(calculatedExtent).toStrictEqual([1, 2, 3, 4]);
});
