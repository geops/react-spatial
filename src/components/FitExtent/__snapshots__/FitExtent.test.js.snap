// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot. 1`] = `
<div
  aria-label="Fit Extent"
  className="tm-button tm-round-blue"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="Fit Extent"
>
  <svg
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 448 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 180V56c0-13.3 10.7-24 24-24h124c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H64v84c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12zM288 44v40c0 6.6 5.4 12 12 12h84v84c0 6.6 5.4 12 12 12h40c6.6 0 12-5.4 12-12V56c0-13.3-10.7-24-24-24H300c-6.6 0-12 5.4-12 12zm148 276h-40c-6.6 0-12 5.4-12 12v84h-84c-6.6 0-12 5.4-12 12v40c0 6.6 5.4 12 12 12h124c13.3 0 24-10.7 24-24V332c0-6.6-5.4-12-12-12zM160 468v-40c0-6.6-5.4-12-12-12H64v-84c0-6.6-5.4-12-12-12H12c-6.6 0-12 5.4-12 12v124c0 13.3 10.7 24 24 24h124c6.6 0 12-5.4 12-12z"
    />
  </svg>
</div>
`;
